### PR TITLE
add garbage collection controller for containers

### LINKED ISSUE #120: [Container Deployer] ContainerDeployer Garbage Collection
**How to categorize this issue?**

/area operations
/kind enhancement
/priority normal

**What would you like to be added**:

Currently the container deployer deploys multiple kubernetes sources in 2 clusters.
e.g. export secrets, state secret, rbac roles, service account, ...

I can happen that the cleanup cannot be correctly performed because of various reasons (manual finalizer deletion etc..)
Therefore, we should have a dedicated periodic garbage collection that removes all container deployer resources where the deployitem does not exist anymore.

### PR DESCRIPTION
**Pull request by schrodit**
**How to categorize this PR?**

/area container-deployer
/kind enhancement
/priority 3

**What this PR does / why we need it**:

Adds a garbage collector for all resources that are created by the container deployer in the host cluster.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/landscaper/issues/120

**Special notes for your reviewer**:

**Release note**:

```feature operator
A GarbageCollector has been added to the Container Deployer to cleanup all resources that are once created by the container deployer.
It is now possible to delete deploy items (and manually remove their finalizer) without a need manually cleanup resources on the host cluster.
```

### CODE REVIEW

**Review by reshnm — CHANGES_REQUESTED** [review]

File: apis/.schemes/container-v1alpha1-Configuration.json [hunk 1]
```diff
@@ -88,6 +88,27 @@
...
+    "container-v1alpha1-GarbageCollection": {
+      "description": "GarbageCollection defines the container deployer garbage collection configuration.",
+      "type": "object",
+      "required": [
+        "disable",
+        "worker"
+      ],
+      "properties": {
+        "disable": {
+          "description": "Disable disables the garbage collector and no resources re cleaned up.",
```
[the remark this answers is not in the document]
Comment (schrodit) [reply apis/.schemes/container-v1alpha1-Configuration.json:100]: oh yes I guess I missed a `a`

File: pkg/deployer/container/garbage_collector.go [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+		if err != nil {
+			return err
+		}
+		gc.log.Info("registered container garbage collector", "resource", "Pods")
+	}
+
+	return nil
+}
+
+func (gc *GarbageCollector) genericResourcesCleanup(obj client.Object) reconcile.Reconciler {
```
[the remark this answers is not in the document]
Comment (schrodit) [reply pkg/deployer/container/garbage_collector.go:116]: it generic in the sense of all rbac related resources.
I renamed it to `cleanupRBACResources`

File: pkg/deployer/container/garbage_collector.go [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+	isLatest, err := gc.isLatestPod(ctx, obj)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if isLatest {
+		logger.V(9).Info("not garbage collected", "reason", "latest pod")
+		return reconcile.Result{}, nil
+	}
+
+	controllerutil.RemoveFinalizer(obj, container.ContainerDeployerFinalizer)
```
[the remark this answers is not in the document]
Comment (schrodit) [reply pkg/deployer/container/garbage_collector.go:206]: reused the already existing `CleanupPod` function

**Review by reshnm — APPROVED** [review]
/lgtm

**Pull request merged**